backend/infrahub/services/adapters/event/__init__.py

```diff
@@ -0,0 +1,46 @@
+from __future__ import annotations
+
+import asyncio
+from typing import TYPE_CHECKING
+
+from prefect.events import emit_event
+
+from infrahub.exceptions import InitializationError
+
+if TYPE_CHECKING:
+    from infrahub.events import InfrahubEvent
+    from infrahub.services import InfrahubServices
+
+
+class InfrahubEventService:
+    """Base class for infrahub event service"""
+
+    def __init__(self) -> None:
+        self._service: InfrahubServices | None = None
+
+    @property
+    def service(self) -> InfrahubServices:
+        if not self._service:
+            raise InitializationError("Event is not initialized with a service")
+
+        return self._service
+
+    async def initialize(self, service: InfrahubServices) -> None:
+        """Initialize the cache"""
+        self._service = service
+
+    async def send(self, event: InfrahubEvent) -> None:
+        tasks = [self._send_bus(event=event), self._send_prefect(event=event)]
+        await asyncio.gather(*tasks)
+
+    async def _send_bus(self, event: InfrahubEvent) -> None:
+        message = event.get_message()
+        await self.service.send(message=message)
+
+    async def _send_prefect(self, event: InfrahubEvent) -> None:
+        emit_event(
+            event=event.get_name(),
+            resource=event.get_resource(),
+            related=event.get_related(),
+            payload=event.get_payload(),
+        )
```

Comment:
This looks a bit problematic, is there an async version of emit_event?

Reply:
Not that I could find quickly but as far as I understand they are sending the event in a non blocking way. I'm planning to dive deeper into it soon.
We may have to create our own function to send the event but for now since we are sending the events in a background task I don't think it will make a big different

Reply:
Ah right the background tasks are run in a separate thread regardless. But we'll have to keep it in mind if used elsewhere, though if Prefects emit_event does the same it should be fine.